tools/meteor-services/deploy.js

```diff
@@ -321,6 +323,206 @@ function canonicalizeSite(site) {
+  let pollingConfiguration = new PollingConfiguration(deployPollTimeout);
+  
+  var deploymentPollResult = buildmessage.enterJob({
+    title: "Waiting for Deploy to succeed..."}, 
+    function() {
+      return pollForDeploy(pollingConfiguration,
+        versionId,
+        true);
+    });
+
+  if(deploymentPollResult && deploymentPollResult.isActive) {
+    // The deploy succeeded and the deployed version is now active
+    const {newVersionNum, hostname, username, region, galaxyUrl} = result.payload;
+    var successfulDeploymentMessage = `
+    ******************************************************************************
+    You have successfully upgraded your app to version ${newVersionNum}.
+    Hostname: ${hostname}, Region: ${region}, Username: ${username}
+    To see the status of your app, visit your app's page on Galaxy
+    \t\t${galaxyUrl}.
+    ******************************************************************************
+    `
+    Console.info(successfulDeploymentMessage);
+  } else {
+    // The deploy did not succeed or the polling timed out
+    if(deploymentPollResult.buildStatus === 'failed-permanently' ||
+      deploymentPollResult.deployStatus === 'failed') {
+        const {newVersionNum, hostname, username, region, galaxyUrl} = result.payload;
+        var failedDeploymentMessage = `
+        ******************************************************************************
+        Failed to deploy your app to version ${newVersionNum}. No changes were made.
+        Hostname: ${hostname}, Region: ${region}, Username: ${username}
+        To see the logs for your app, visit ${galaxyUrl}/logs.
+        ******************************************************************************
+        `
+        Console.info(successfulDeploymentMessage);
+      } else {
+        // The status was non-terminal, so we most likely timed out
+        Console.info(result.payload.message);
+        return 1;
+      }
+  }
+}
+
+// Creates a polling configuration with defaults if fields left unset
+// Default initialWaitTime is 10 seconds – this is the time to wait before checking at all
+// Default startInterval is 700 milliseconds – this is the first wait interval between polls
+// Default backoffFactor is 1.3
+// Default timeout is 15 minutes
+function PollingConfiguration(
+  timeoutMillis,
+  initialWaitTimeMillis,
+  startIntervalMillis,
+  backoffFactor) {
+    const fifteenMinutesMillis = 15*60*1000;
+    this.initialWaitTimeMillis = initialWaitTimeMillis || 10*1000;
+    this.startIntervalMillis = startIntervalMillis || 700;
+    this.backoffFactor = backoffFactor || 1.3;
+    this.timeoutMillis = timeoutMillis || fifteenMinutesMillis;
+  }
+
+// Build statuses that indicate we should stop polling for build
+const terminalBuildStatuses = ['failed-permanently', 'success']
+// Deploy statuses that indicate we should stop polling for deploy
+const terminalDeployStatuses = ['success', 'failed', 'aborted']
```

Comment:
See related comments about "terminal" below... terminology wise, this seems to be less like a tree with children which might be terminal and more like a recursive.  Maybe this can just be `finishedBuildStatuses` and `finishedDeployStatuses`?

Reply:
FWIW 'terminal' makes sense to me but if it doesn't to you, that seems like a reasonable reason to change.

Though I wonder if we should actually allow ourselves some future flexibility by just adding another field to version-status that's a bool "finished", so if we add another state later we don't break old clients. Like ideally we could make all of the state machine logic here driven by bools rather than recognizing lists of states (so same for the build -> deploy transition).

Reply:
So @glasser are you suggesting moving this logic to galaxy-server? I think that could make a lot of sense. I'll put out a PR for galaxy-server that does that 👍